### ISSUE TITLE
os-key level python support (osx)

### ISSUE THREAD

**Issue by Thetoxicarcade** [travis-ci/travis-ci#7132]
As mentioned in the docs and on issue #2312, Python isn't quite supported on OSX. Running it, however, it does appear to be trying to do something with ruby/python when the matrix is set to do it (see [example error](https://travis-ci.org/congredi/congredi/jobs/190559253)).

```
$ rvm use
Warning! PATH is not properly set up, '/Users/travis/.rvm/gems/ruby-2.0.0-p648/bin' is not at first place,
         usually this is caused by shell initialization files - check them for 'PATH=...' entries,
         it might also help to re-add RVM to your dotfiles: 'rvm get stable --auto-dotfiles',
         to fix temporarily in this shell session run: 'rvm use ruby-2.0.0-p648'.
Using /Users/travis/.rvm/gems/ruby-2.0.0-p648
2.7 is not installed; attempting download
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.
The command "sudo tar xjf python-2.7.tar.bz2 --directory /" failed and exited with 1 during .
Your build has been stopped.
/Users/travis/build.sh: line 148: shell_session_update: command not found
```

Current uses of osx with python use a workaround, like:

```
matrix:
  include:
    - os: osx
      sudo: required
      language: generic
```
But that requires knowing what version of python you're testing, as `TRAVIS_PYTHON_VERSION` is unset:
```
before_install:
  # give this a spin
  - if [[ $TRAVIS_OS_NAME == 'osx' ]]; then brew update; fi
  - if [[ $TRAVIS_OS_NAME == 'osx' ]]; then brew outdated python || brew upgrade python; fi
  # generic won't set TRAVIS_PYTHON_VERSION
  - if [[ $TRAVIS_OS_NAME == 'osx' && $TRAVIS_PYTHON_VERSION == 3* ]]; then brew install python3; fi
  - if [[ $TRAVIS_OS_NAME == 'osx' && $TRAVIS_PYTHON_VERSION == '2.7' ]]; then brew outdated python2.7 || brew upgrade python2.7; fi
```

In short, do we support the OSX key for python yet?
```
language: python
os:
  - osx
  - linux
```

**Comment by BanzaiMan** [comment]
I don't think there is an issue here.

With `os: osx` and `language: python`, the job is correctly routed to the Mac queue. We run `rvm use` on OS X because of https://github.com/travis-ci/travis-ci/issues/6674.

The build halts with the message:

```
2.7 is not installed; attempting download
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.
```

when the build script tries to download a Python archive for OS X, but is unable to find it.

Fixing this issue basically realizes the Python support.

The need for testing `$TRAVIS_OS_NAME` and `$TRAVIS_PYTHON_VERSION` is not dependent on `os` key support (because this part is functioning fine, as explained above). The need is due to the Python runtime not being set up correctly when the build gets to it (so you are installing correct one with Homebrew), which is precisely what #2312 addresses.

In summary, everything this issue raises is either invalid, or addressed by #2312. As such, I am closing this as a duplicate.

**Comment by Thetoxicarcade** [comment]
So you're saying
```
language: python
os:
 - osx
 - linux
```
is valid, instead of the two (possibly 4, depending on how you look at it) workarounds on #2312? Workarounds don't solve the problem. Can the default travis worker, on osx, use different python versions without a complicated/verbose workaround?

#6674 addresses running an osx build on `language: generic`, which is **a workaround**, not a solution to the problem. To make language: generic work, you have to run `brew` in the `before_install` section to get the virtualenv in place. That's not the same as solving the actual problem.

That's not really addressing the problem if it doesn't work. If it does, can we update the docs that say this limitation still exists? [multi os](https://docs.travis-ci.com/user/multi-os/), [osx build environment](https://docs.travis-ci.com/user/osx-ci-environment/#Languages)

You can't say it supports python unless it supports it **the same as** the linux builds, because otherwise we go through a confusing process to setup python ourselves, instead of letting travis-ci's build.sh start a virtualenv for us.

**Comment by BanzaiMan** [comment]
The workarounds suggested in #2312 are just that: workarounds. They do not realize the support entirely.

> Can the default travis worker, on osx, use different python versions without a complicated/verbose workaround?

The answer to this question is "yes", if the archives are available. But the build currently does not get to that, so the question itself is moot.

> #6674…

I did not say this has anything to do with Python-on-Mac. I brought it up only to indicate that its presence does not mean we are treating the `os: osx`-`language: python` jobs as Ruby jobs.

As I said, we do not yet support Python on the Mac. And the issues are discussed in #2312.

**Comment by Thetoxicarcade** [comment]
Thank you. So we don't, is the answer I wanted.

**Comment by BanzaiMan** [comment]
"We don't", what?
